sunpy/io/file_tools.py

```diff
@@ -77,18 +79,28 @@ def read_file(filepath, filetype=None, **kwargs):
     -----
     Other keyword arguments are passed to the reader used.
     """
+    if not isinstance(filepath, (str, io.IOBase, pathlib.Path)):
+        raise TypeError("expected str, IO or pathlib.Path object")
+
+    if isinstance(filepath, io.IOBase):
+        _filepath = pathlib.Path(filepath.name)
```

Comment:
Is there a reason for the underscore?

Reply:
feel free to correct me, but doesn't the detection logic requires str filepath?
https://github.com/sunpy/sunpy/blob/0f7946d90dae9eeb6ae8e81913a32e0745ee7b44/sunpy/io/file_tools.py#L102-L103

Reply:
on looking closely, I feel like making the distinction there is a better way; we'll have to extract filetype anyway for,
https://github.com/sunpy/sunpy/blob/0f7946d90dae9eeb6ae8e81913a32e0745ee7b44/sunpy/io/file_tools.py#L97-L99
plus it appears unnecessary to add compatibiliy/support for file-handlers in the detection logic.


ps - looks like there are no tests for the detection function, shall I add one? https://github.com/sunpy/sunpy/blob/0f7946d90dae9eeb6ae8e81913a32e0745ee7b44/sunpy/io/file_tools.py#L196

Reply:
Sure add a test for `detect_filetype`. 

I am not sure I follow on making the distinction, what do you want to change?

Reply:
I think the detection works by reading some lines from the file, so its probably better to have a separate file obj in the detection function(current implementation) to avoid unnecessary bugs and ResourceWarnings. (probably bugs like unintentionally closing original file-obj, etc)

Reply:
Ok.